Exit batch files explictly using ERRORLEVEL

Mpdreamz opened the pull request:
This makes sure the exit code is preserved when calling the batch
files from different contexts other than DOS

Fixes #29582

Mpdreamz committed:
Exit batch files explictly using ERRORLEVEL

This makes sure the exit code is preserved when calling the batch
files from different contexts other than DOS

Fixes #29582

This also fixes specific error codes being masked by an explict

exit /b 1

causing the useful exitcodes from ExitCodes to be lost.

elasticmachine commented:
Pinging @elastic/es-core-infra

rjernst commented:
@Mpdreamz Do you want to bring this PR up to date? There are a number of new bat files now that xpack has been opened.

Mpdreamz commented:
I've updated this PR and also fixed cases where we were always setting the exit code to `1` in case of an error.

e.g currently `bin\elasticsearch-shard.bat` always returns `1` while the cli tool itself is returning the more useful [`64` exit code](https://github.com/elastic/elasticsearch/blob/master/libs/cli/src/main/java/org/elasticsearch/cli/ExitCodes.java#L25) 


cc @elastic/es-core-infra @elastic/microsoft

jasontedor commented on distribution/src/bin/elasticsearch-keystore.bat:
```diff
@@ -4,9 +4,8 @@ setlocal enabledelayedexpansion
-call "%~dp0elasticsearch-cli.bat" ^
-  %%* ^
-  || exit /b 1
+call "%~dp0elasticsearch-cli.bat" %%* 
```
Is there a reason this wouldn't work? Then we can keep the exit close to the failing script, we really do want to error out immediately here.

```suggestion
call "%~dp0elasticsearch-cli.bat" ^
  %%* ^
  || exit /b %errorlevel%
```

Mpdreamz replied:
Thanks for the review @jasontedor, finally took the effort to get back to this PR.

Sadly we need the `exit /b %errorlevel%` after the `endlocal` calls otherwise it does not survive in a non `CMD` context.

I also tried the special `:eof` label as [documented here](https://blogs.msdn.microsoft.com/oldnewthing/20120802-00/?p=6973)


```DOS
call "%~dp0elasticsearch-cli.bat" %%* || goto :eof
```

We could get rid of the `endlocal` calls alltogether since the local context is automatically closed when the bat file exits but non CMD contexts still need the explicit `exit /B`

I will update the PR with the following:

```bat
call "%~dp0elasticsearch-cli.bat" %%* || goto exit

endlocal
endlocal
:exit
exit /B %ERRORLEVEL%
```

Which I think makes the early exit explicit and satisfies carrying over the exit code to non CMD contexts.

jasontedor commented on distribution/src/bin/elasticsearch-keystore.bat:
```diff
@@ -4,9 +4,8 @@ setlocal enabledelayedexpansion
 
 endlocal
 endlocal
+exit /b %ERRORLEVEL%
```
```suggestion
```

jasontedor reviewed (commented):
I left a question.

jasontedor commented on distribution/src/bin/elasticsearch-keystore.bat:
```diff
@@ -4,9 +4,9 @@ setlocal enabledelayedexpansion
-call "%~dp0elasticsearch-cli.bat" ^
-  %%* ^
-  || exit /b 1
+call "%~dp0elasticsearch-cli.bat" %%* || goto exit
```
Would you maintain the line breaks that we have? It keeps these scripts looking the same as the Unix scripts too.

jasontedor reviewed (changes requested):
I left one request, about the line breaks.

Mpdreamz committed:
fix line breaks for calling cli to match the bash scripts

Mpdreamz committed:
indent size of bash files is 2, make sure editorconfig does the same for bat files

Mpdreamz committed:
update indenting to match bash files

Mpdreamz committed:
update elasticsearch-keystore.bat indenting

Mpdreamz commented:
@jasontedor, updated to the linebreaks on the bat files to match the unix scripts.

Mpdreamz commented:
Test this please

jasontedor reviewed (changes requested):
There are two new scripts:
 - elasticsearch-node
 - elasticsearch-shard

Would you address those as well please? Otherwise this looks good. Thanks for picking this up.

Mpdreamz committed:
Update elasticsearch-node.bat to exit outside of endlocal

jasontedor reviewed (approved):
LGTM.
